File: Sources/Application/ReconnectBackoff.swift
struct ReconnectBackoff: Equatable, Sendable {
  private static let maximumDelay: UInt64 = 30
  private var nextValue: UInt64 = 1

  mutating func nextDelaySeconds() -> UInt64 {
    let current = nextValue
    nextValue = min(nextValue * 2, Self.maximumDelay)
    return current
  }

  mutating func reset() {
    nextValue = 1
  }
}

File: platform/windows/MihomoMeter.Windows.Core/Application/ReconnectBackoff.cs
namespace MihomoMeter.Windows.Core.Application;

public sealed class ReconnectBackoff
{
    private const int MaximumDelaySeconds = 30;
    private int _nextValue = 1;

    public int NextDelaySeconds()
    {
        var current = _nextValue;
        _nextValue = Math.Min(_nextValue * 2, MaximumDelaySeconds);
        return current;
    }

    public void Reset()
    {
        _nextValue = 1;
    }
}
